### PR TITLE
fix split routes route definition (issue #25)

### PR DESCRIPTION
**Pull request by Valantin**
this fix the route to use ppp ip as gateway for vpn subnet.
need to add the definition of a static route to reach the gateway if the gateway subnet match vpn subnet

### CODE REVIEW

**Comment by adrienverge** [comment]
I'm sorry but I don't understand what this pull request does.

The commit message needs to explain what it does, the reason why, and if we should expect any change for other users.

Thanks!

**Comment by Valantin** [comment]
Issue #25 

the ip assigned by fortigate to my client is 10.30.0.5, the subnet to add route is 172.16.0.0/20

official client (working)
ip route
10.10.152.0/22 via 10.30.0.5 dev ppp0
172.16.0.0/20 via 10.30.0.5 dev ppp0

openfortivpn (not working)
ip route
10.10.152.0/22 dev ppp0 scope link
172.16.0.0/20 dev ppp0 scope link

openfortivpn patched (working)
ip r
10.10.152.0/22 via 10.30.0.5 dev ppp0
172.16.0.0/20 via 10.30.0.5 dev ppp0 


My code add flag RTF_GATEWAY to add "via" in the route table, check if "s_addr == 0" (not -1, is unsigned int) to add "10.30.0.5" if no gateway set

no change expected for other user.

**Comment by mrbaseman** [comment]
I have created a testing branch and at a first glance this patch doesn't break things in my setup

**Comment by adrienverge** [comment]
I'm sorry but I still don't understand what this does: is it a bug fix? If yes, what's the bug? How does it solve it? Please provide a *clear* explanation. And the commit message needs to contain this explanation.

**Comment by mrbaseman** [comment]
actually, for me the "via" part is added to the routes without this patch already. @Valantin which versions of openfortivpn did you compare? A quite recent checkout from master with and without the proposed patch?

File: src/ipv4.c [hunk 1]
```diff
@@ -281,8 +281,9 @@ static int ipv4_set_split_routes(struct tunnel *tunnel)
 		route = &tunnel->ipv4.split_rt[i];
 		strncpy(route_iface(route), tunnel->ppp_iface,
 		        ROUTE_IFACE_LEN - 1);
-		if (route_gtw(route).s_addr == -1)
+		if (route_gtw(route).s_addr == 0)
```
Comment (mrbaseman) [inline src/ipv4.c]: This is the decision if the route has already a gateway set or not. -1 casted to an unsigned int would result in UINT_MAX which is 255.255.255.255 in IP notation. Routes without a gateway however carry the gateway IP 0.0.0.0 which is the proposed 0 here. With -1 the following line would (normally) never be executed. So this is a bugfix.

File: src/ipv4.c [hunk 2]
```diff
@@ -281,8 +281,9 @@ static int ipv4_set_split_routes(struct tunnel *tunnel)
...
 			route_gtw(route).s_addr = tunnel->ipv4.ip_addr.s_addr;
+		route->rt_flags |= RTF_GATEWAY;
```
Comment (mrbaseman) [inline src/ipv4.c]: this adds the gateway flag to the routes received from the Fortigate. Without the gateway flag the destination is directly connected (over the vpn tunnel), which is not always the case. There might be more hops behind the remote gateway to reach the network to which this route points.

**Comment by Valantin** [comment]
@mrbaseman I've tested the master branch with a fortigate 800c with fortios 5.0
My VPN is configured to use a /16 subnet instead the default ip pool for client connection

@adrienverge sorry I can't explain better the problem.
In my setup this partially fix issue #25 the problem is described in the issue

**Comment by mrbaseman** [comment]
@Valantin: Thanks for your reply. In the meantime I understand better what these changes are good for (see my comments on the changes).
@adrienverge: Does this make sense to you? As you can see, I have also pulled these changes into my other work on the routing code

**Comment by Valantin** [comment]
@mrbaseman I've tried your ppp-routes branch, It seems to fix issue #25 
On my setup, fortigate gateway is in the same /23 routed by fortigate vpn, No more need to add manualy the /32 route

**Comment by adrienverge** [comment]
@mrbaseman thanks for taking the time to explain this change.

@Valantin: as said before
> The commit message needs to explain what it does, the reason why, and if we should expect any change for other users.

> the commit message needs to contain this explanation.

it is important to be able to track changes and `git blame` problems using the git history (and impendently of GitHub).

**Comment by mrbaseman** [comment]
now that #95 has been merged this one can be closed. I have picked up the proposed changes in the larger rework of the routing code.

**Pull request closed without merging**